BorrowBookControl.java

```diff
@@ -2,116 +2,116 @@
-		}	
-		for (book B : PENDING) {
-			loan LOAN = LIBRARY.ISSUE_LAON(B, M);
-			COMPLETED.add(LOAN);			
-		}
-		UI.Display("Completed Loan Slip");
-		for (loan LOAN : COMPLETED) {
-			UI.Display(LOAN.toString());
-		}
-		UI.Set_State(BorrowBookUI.UI_STATE.COMPLETED);
-		State = CONTROL_STATE.COMPLETED;
-	}
-
-	
-	public void cancel() {
-		UI.Set_State(BorrowBookUI.UI_STATE.CANCELLED);
-		State = CONTROL_STATE.CANCELLED;
-	}
-	
-	
+
+    private BorrowBookUI UI;
+
+    private library library;
+    private member member;
+    private enum ControlState { INITIALISED, READY, RESTRICTED, SCANNING, IDENTIFIED, FINALISING, COMPLETED, CANCELLED };
+    private ControlState state;
+
+    private List<book> pending;
+    private List<loan> completed;
+    private book book;
+
+
+    public BorrowBookControl() {
+        this.library = library.INSTANCE();
+        this.state = ControlState.INITIALISED;
+    }
+
+
+    public void setUI(BorrowBookUI UI) {
+        if (!this.state.equals(ControlState.INITIALISED))
+            throw new RuntimeException("BorrowBookControl: cannot call setUI except in INITIALISED state");
+
+        this.UI = UI;
+        UI.setState(BorrowBookUI.UIState.READY);
+        this.state = ControlState.READY;
+    }
+
+
+    public void swiped(int memberId) {
+        if (!this.state.equals(ControlState.READY))
+            throw new RuntimeException("BorrowBookControl: cannot call cardSwiped except in READY state");
+
+        this.member = library.MEMBER(memberId);
+        if (this.member == null) {
+            this.UI.display("Invalid memberId");
+            return;
+        }
+        if (this.library.MEMBER_CAN_BORROW(this.member)) {
+            this.pending = new ArrayList<>();
+            this.UI.setState(BorrowBookUI.UIState.SCANNING);
+            this.state = ControlState.SCANNING; }
+        else
+        {
+            this.UI.display("Member cannot borrow at this time");
+            this.UI.setState(BorrowBookUI.UIState.RESTRICTED); }}
+
+
+    public void scanned(int bookId) {
+        this.book = null;
+        if (!this.state.equals(ControlState.SCANNING)) {
+            throw new RuntimeException("BorrowBookControl: cannot call bookScanned except in SCANNING state");
+        }
+        this.book = this.library.Book(bookId);
+        if (this.book == null) {
+            this.UI.display("Invalid bookId");
+            return;
+        }
+        if (!this.book.AVAILABLE()) {
+            this.UI.display("Book cannot be borrowed");
+            return;
+        }
+        this.pending.add(this.book);
+        for (book displayBook : this.pending) {
+            this.UI.display(displayBook.toString());
+        }
+        if (this.library.Loans_Remaining_For_Member(this.member) - this.pending.size() == 0) {
+            this.UI.display("Loan limit reached");
+            this.complete();
+        }
+    }
+
+
+    public void complete() {

```

Comment:
D8: use a method name more meaningful than 'complete'

Reply:
updated to borrowBookComplete

Reply:
D8.1: please use the verbNoun format to be clear. Method 'borrowBookComplete' could be mistaken as a boolean variable.
e.g. completeBorrow

Reply:
done